## Team Members

<p>VAIBHAV     2021H1540846P</p>
PRAJESH RAO 2021H1540861P

In [0]:
import pyspark
from pyspark import SparkFiles
from pyspark.sql.functions import to_timestamp
from pyspark.sql.functions import to_date
from pyspark.sql.functions import randn #For generating dataframe 
from pyspark.sql.functions import mean as _mean, stddev as _stddev, col

spark.sql("set spark.sql.legacy.timeParserPolicy=LEGACY")
import pandas as pd
import numpy as np

<h1><b>Problem 1</b></h1>
<H4>
<p>1. Load “linearRegressionData.csv” to Databricks environment.</p>
<p>2. Create a dataframe with 500k simulated data records from “linearRegressionData.csv”</p>
<p>3. Load “autoMPGDataModified.csv” to Databricks environment</p>
<p>4. Create a dataframe with 500k simulated data records from “autoMPGDataModified.csv”</p>

In [0]:
# File location and type
file_location1 = "/FileStore/tables/autoMPGDataModified-1.csv"
file_location2 = "/FileStore/tables/linearRegressionData-1.csv"
file_type = "csv"

# CSV options
infer_schema = "true"
first_row_is_header = "true"
delimiter = ","

# The applied options are for CSV files. For other file types, these will be ignored.
autompg = spark.read.format(file_type).option("inferSchema", infer_schema) .option("header", first_row_is_header).option("sep", delimiter) .load(file_location1)
linerregression = spark.read.format(file_type).option("inferSchema", infer_schema).option("header", first_row_is_header).option("sep", delimiter).load(file_location2)

display(autompg)
display(linerregression)

mpg displacement horsepower weight accelaration 18 307 18 3504 12.0 15 350 36 3693 11.5 18 318 30 3436 11.0 16 304 30 3433 12.0 17 302 25 3449 10.5

dvs1 ivs1 ivs2 ivs3 34.63 5.53 5.58 5.41 40.89 3.89 6.48 6.97 37.25 5.07 4.5 6.5 45.09 5.81 5.71 8.59 39.4 5.61 5.79 6.77

In [0]:
# Checking Schema of thre DataFrames
autompg.printSchema()
linerregression.printSchema()

root
 |-- mpg: integer (nullable = true)
 |-- displacement: integer (nullable = true)
 |-- horsepower: integer (nullable = true)
 |-- weight: integer (nullable = true)
 |-- accelaration: double (nullable = true)

root
 |-- dvs1: double (nullable = true)
 |-- ivs1: double (nullable = true)
 |-- ivs2: double (nullable = true)
 |-- ivs3: double (nullable = true)



In [0]:
# Coverting RDD to dataframe
autompg=autompg.toPandas()
linerregression=linerregression.toPandas()

<b>Simulating the values for each column  based on given dataset using random function

In [0]:

N = 500000   

# For autompg.csv
mpgmean = autompg.agg({'mpg': 'mean'})
mpgstd = autompg['mpg'].std()
random_mpgmean = np.random.normal(loc=mpgmean, scale=mpgstd, size=N)

displacementmean = autompg.agg({'displacement': 'mean'})
displacementstd = autompg['displacement'].std()
random_displacement = np.random.normal(loc=mpgmean, scale=mpgstd, size=N)

horsepowermean = autompg.agg({'horsepower': 'mean'})
horsepowerstd = autompg['horsepower'].std()
random_horsepower = np.random.normal(loc=mpgmean, scale=mpgstd, size=N)
 
weightmean = autompg.agg({'weight': 'mean'})
weightstd = autompg['weight'].std()
random_weight = np.random.normal(loc=mpgmean, scale=mpgstd, size=N)
 
accelarationmean = autompg.agg({'accelaration': 'mean'})
accelarationstd = autompg['accelaration'].std()
random_acceleration = np.random.normal(loc=accelarationmean, scale=accelarationstd, size=N)

# For linearregression.csv
dvs1mean = linerregression.agg({'dvs1': 'mean'})
dvs1std = linerregression['dvs1'].std()
random_dvs1mean = np.random.normal(loc=dvs1mean, scale=dvs1std, size=N)

ivs1mean = linerregression.agg({'ivs1': 'mean'})
ivs1std = linerregression['ivs1'].std()
random_ivs1mean = np.random.normal(loc=ivs1mean, scale=ivs1std, size=N)

ivs2mean = linerregression.agg({'ivs2': 'mean'})
ivs2std = linerregression['ivs2'].std()
random_ivs2mean = np.random.normal(loc=ivs2mean, scale=ivs2std, size=N)
 
ivs3mean = linerregression.agg({'ivs3': 'mean'})
ivs3std = linerregression['ivs3'].std()
random_ivs3mean = np.random.normal(loc=ivs3mean, scale=ivs3std, size=N)

<b> Creating a DataFrame

In [0]:
# For autompg
autompg = pd.DataFrame()
autompg['mpg'] = random_mpgmean.tolist()
autompg['displacement'] = random_displacement.tolist()
autompg['horsepower'] = random_horsepower.tolist()
autompg['weight'] = random_weight.tolist()
autompg['accelaration'] = random_acceleration.tolist()

# For linearregression
linerregression = pd.DataFrame()
linerregression['dvs1'] = random_dvs1mean.tolist()
linerregression['ivs1'] = random_ivs1mean.tolist()
linerregression['ivs2'] = random_ivs2mean.tolist()
linerregression['ivs3'] = random_ivs3mean.tolist()

In [0]:
linerregression

dvs1 
 ivs1 
 ivs2 
 ivs3 
 
 
 
 
 0 
 30.868607 
 4.906146 
 5.653928 
 8.000984 
 
 
 1 
 30.877046 
 5.115470 
 5.652047 
 6.963801 
 
 
 2 
 39.960180 
 3.655901 
 5.492938 
 7.678221 
 
 
 3 
 41.172081 
 4.627264 
 6.299474 
 5.146269 
 
 
 4 
 46.428122 
 5.897832 
 4.929974 
 8.746798 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 499995 
 36.269705 
 3.934434 
 4.793681 
 6.617237 
 
 
 499996 
 35.319481 
 3.895673 
 6.264085 
 7.422195 
 
 
 499997 
 42.841171 
 5.038462 
 5.872625 
 8.211394 
 
 
 499998 
 42.646160 
 4.195027 
 6.863068 
 6.740368 
 
 
 499999 
 41.459136 
 3.743306 
 5.758783 
 5.269588 
 
 
 
 500000 rows × 4 columns

In [0]:
autompg

mpg 
 displacement 
 horsepower 
 weight 
 accelaration 
 
 
 
 
 0 
 16.214635 
 16.071718 
 14.587797 
 15.449113 
 11.601432 
 
 
 1 
 18.543724 
 17.078059 
 18.573899 
 20.828405 
 12.750307 
 
 
 2 
 16.777088 
 15.840512 
 17.734115 
 16.478783 
 12.351655 
 
 
 3 
 18.745539 
 17.471584 
 18.825469 
 16.890875 
 11.359075 
 
 
 4 
 16.267082 
 18.232842 
 19.373804 
 17.644225 
 11.244740 
 
 
 ... 
 ... 
 ... 
 ... 
 ... 
 ... 
 
 
 499995 
 18.151181 
 19.316755 
 19.651393 
 14.116581 
 11.637874 
 
 
 499996 
 19.582389 
 14.989716 
 18.863596 
 16.517724 
 11.296449 
 
 
 499997 
 16.998015 
 16.844012 
 15.126942 
 16.130710 
 11.697808 
 
 
 499998 
 16.475954 
 17.823013 
 17.671518 
 15.939338 
 12.075105 
 
 
 499999 
 17.565164 
 16.734608 
 16.315455 
 16.918353 
 12.533589 
 
 
 
 500000 rows × 5 columns

<h1><b>Problem 2</b></h1>
<H4>
<p>1. For data “linearRegressionData.csv”, transform Dataframe  to RDD.</p>
<p>2. Create an RDD of the labeled point.</p>
<p>3. Divide the data into a training and testing set.</p>
<p>4. Create a linear regression model.</p>
<p>5. Train and save the model.</p>
<p>6. Predict data using the saved model.</p>
<p>7. Evaluate the created model and check its accuracy.</p>

In [0]:
from pyspark.mllib.regression import LabeledPoint
import pyspark.mllib
import pyspark.mllib.regression
from pyspark.sql.functions import *
import numpy as np
from pyspark.mllib.linalg import Vectors

# Created RDD FILE of linearregression , used for prediction without labeled point
linerregression_f = spark.createDataFrame(linerregression)

# Used for creating Labeled point and Prediction
linerregression_frdd=linerregression_f.rdd
print("RDD File of linear regression")
linerregression_frdd.take(5)

RDD File of linear regression
Out[186]: [Row(dvs1=30.868606873496397, ivs1=4.906146088675779, ivs2=5.653927772962623, ivs3=8.000983945097717),
 Row(dvs1=30.87704626606618, ivs1=5.115470343336037, ivs2=5.6520469271794544, ivs3=6.9638010871957485),
 Row(dvs1=39.96018042758738, ivs1=3.6559007274062254, ivs2=5.492938121094742, ivs3=7.678220894244441),
 Row(dvs1=41.172080581335415, ivs1=4.627264365252115, ivs2=6.2994737534263585, ivs3=5.146268588358932),
 Row(dvs1=46.428122083787386, ivs1=5.897831609743697, ivs2=4.929973595212521, ivs3=8.746797757644341)]

<center> <h2> Without Labeled Point</h2></center>

In [0]:
# Vectorization of Data of linearregression
from pyspark.ml.feature import VectorAssembler
linerregression_f = linerregression_f.withColumnRenamed('dvs1', "target")  

vectorAssembler = VectorAssembler(inputCols = ['ivs1','ivs2','ivs3',], outputCol = 'features')
linerregression_df_assembled = vectorAssembler.transform(linerregression_f)
linerregression_df_assembled = linerregression_df_assembled.select(['features', 'target'])
linerregression_df_assembled.show(5, truncate=False)

+---------------------------------------------------------+------------------+
|features                                                 |target            |
+---------------------------------------------------------+------------------+
|[4.906146088675779,5.653927772962623,8.000983945097717]  |30.868606873496397|
|[5.115470343336037,5.6520469271794544,6.9638010871957485]|30.87704626606618 |
|[3.6559007274062254,5.492938121094742,7.678220894244441] |39.96018042758738 |
|[4.627264365252115,6.2994737534263585,5.146268588358932] |41.172080581335415|
|[5.897831609743697,4.929973595212521,8.746797757644341]  |46.428122083787386|
+---------------------------------------------------------+------------------+
only showing top 5 rows



In [0]:
# Dividing the Dataset into training and test data
(train_data, test_data) = linerregression_df_assembled.randomSplit([0.7, 0.3])
print("Train count:", train_data.count())
print("Test count:", test_data.count())

Train count: 350409
Test count: 149591


In [0]:
# Developing a Linear Regression Model 
from pyspark.ml.regression import LinearRegression
rf1 = LinearRegression(featuresCol="features", labelCol="target", maxIter=100,regParam=0.0, elasticNetParam=0.0,aggregationDepth=2,fitIntercept=True, standardization=True)

# Training the Data
modellr = rf1.fit(train_data)
print("Coefficients: " + str(modellr.coefficients))
print("Intercept: " + str(modellr.intercept))
print("")
print("Training Summary")

trainingSummary = modellr.summary
print("RMSE: %f" % trainingSummary.rootMeanSquaredError)
print("r2: %f" % trainingSummary.r2)
print("objectiveHistory: %s" % str(trainingSummary.objectiveHistory))
trainingSummary.residuals.show()

Coefficients: [-0.0045587878969564856,0.002371827008682636,0.005382462584030815]
Intercept: 39.4293678517764

Training Summary
RMSE: 3.934384
r2: 0.000003
objectiveHistory: [0.0]
/databricks/spark/python/pyspark/sql/context.py:134: FutureWarning: Deprecated in 3.0.0. Use SparkSession.builder.getOrCreate() instead.
  warnings.warn(
+--------------------+
|           residuals|
+--------------------+
|   2.800959136192745|
|  2.2157652623831723|
|   0.736872657476006|
| -6.6524385809673845|
|  -0.713645908347118|
|   7.035896510589687|
|  1.1892097538492195|
|  1.5384680882085675|
|-0.15524639953208919|
|   9.166667920618664|
| -1.2137078893998208|
|  -9.302047852203778|
|   5.667437349420737|
|  -8.190345121318273|
|    7.57836241648031|
|  0.7973722106471044|
|  0.8023629456414341|
|  0.8066814644404801|
| -4.4877359437485325|
|  0.1601572022551423|
+--------------------+
only showing top 20 rows



In [0]:
# For Deleteing  an Existing Path
dbutils.fs.rm('/tmp/mllib-persistence-example/modellr', True)

# Saving an Model
from pyspark.ml.regression import LinearRegression

# let modellr be a Regression Model
basePath = "/tmp/mllib-persistence-example"
modellr.save(basePath + "/modellr")

# Loading an Model
from pyspark.ml.regression import LinearRegressionModel
loaded_model_lr=LinearRegressionModel.load('/tmp/mllib-persistence-example//modellr')

In [0]:
#Prediction on Testing Data using Loaded Model and Evaluation
from pyspark.ml.evaluation import RegressionEvaluator
evaluator = RegressionEvaluator(labelCol="target", predictionCol="prediction", metricName="rmse")

predictions_lr = loaded_model_lr.transform(test_data)

predictions_lr.select("features","target", "prediction").show(10)
rmse1 = evaluator.evaluate(predictions_lr)
print("Linear Regression RMSE:", rmse1)


+--------------------+------------------+------------------+
|            features|            target|        prediction|
+--------------------+------------------+------------------+
|[2.01309984572106...| 41.68149210005434| 39.46638732852593|
|[2.23838755062637...|38.882042292211835| 39.47159304864827|
|[2.26875087286585...|42.247012457563734|39.464031769126755|
|[2.27117168431004...|40.141087310279545| 39.47875801522439|
|[2.38318675313264...| 36.67953122131718| 39.46596101018731|
|[2.38390783611457...|  25.7613157222561| 39.47194136870818|
|[2.50542177912622...| 39.76583651431153| 39.46030058626489|
|[2.54763963887989...| 40.90935335463097| 39.46991094583617|
|[2.56452197414693...| 38.86840888686479| 39.45414208180543|
|[2.59993252573539...|  34.7026223227585| 39.47135159452523|
+--------------------+------------------+------------------+
only showing top 10 rows

Linear Regression RMSE: 3.930262331214548


<h2>With Labeled Point</h2>

In [0]:
#test train split
split = linerregression_frdd.randomSplit([0.7,0.3],seed=1)
train_data = split[0]
test_data = split[1]

In [0]:
from pyspark.mllib.regression import LabeledPoint

Labeled_Point_train = train_data.map(lambda data : LabeledPoint(data[0],data[1:4]))
Labeled_Point_test = test_data.map(lambda data : LabeledPoint(data[0],data[1:4]))

In [0]:
from pyspark.mllib.regression import LinearRegressionWithSGD as sgd
Linear_Regression_LP = sgd.train(data = Labeled_Point_train,iterations = 100,step = 0.05,intercept = True)

/databricks/spark/python/pyspark/mllib/regression.py:301: FutureWarning: Deprecated in 2.0.0. Use ml.regression.LinearRegression.
  warnings.warn(


In [0]:
print("Intercept:")
print(Linear_Regression_LP.intercept)
print("Coefficients of the Variables:")
print("ivs1:",Linear_Regression_LP.weights[0])
print("ivs2:",Linear_Regression_LP.weights[1])
print("ivs3:",Linear_Regression_LP.weights[2])

Intercept:
1.5976868756528104
Coefficients of the Variables:
ivs1: 2.140972314486033
ivs2: 1.865477973302736
ivs3: 1.865477973302736


In [0]:
# For Deleteing  an Existing Path
dbutils.fs.rm('/home/Linear_Regression_LP', True)

# Saving an Model
from pyspark.ml.regression import LinearRegression
Linear_Regression_LP.save(sc, '/home/Linear_Regression_LP')

# Loading an Model
from pyspark.mllib.regression import LinearRegressionModel as LinearModel
Loaded_LP_LR = LinearModel.load(sc, '/home/Linear_Regression_LP')

# Predicting
predicted = Labeled_Point_test.map(lambda data : (float(data.label) , float(Loaded_LP_LR.predict(data.features))))

In [0]:
predicted.take(5)

Out[197]: [(30.87704626606618, 39.347562729889674),
 (46.428122083787386, 42.5848166740473),
 (39.23365542878202, 36.36104601454085),
 (38.84937342074121, 35.70631924614586),
 (39.53962528985326, 32.07323979794679)]

In [0]:
#RMSE Value
from pyspark.mllib.evaluation import RegressionMetrics as metrics
error = metrics(predicted)
print("RMSE:")
print(error.rootMeanSquaredError)

# R squared value
print("R Squared Value")
print(error.r2)

RMSE:
5.084526365009351
R Squared Value
-1.5022670735377255


<h1><p>Problem 3</h1></p>
<H4>
<p> Perform the following operations</p>
<p>1. For data “autoMPGDataModified.csv”, transform Dataframe to RDD.</p>
<p>2. Create an RDD of the labeled point.</p>
<p>3. Divide the data into a training and testing set.</p>
<p>4. Create a ridge regression model.</p>
<p>5. Train and save the model.</p>
<p>6. Predict data using the saved model.</p>
<p>7. Evaluate the created model and check its accuracy.</p>
<p>8. Create a lasso regression model.</p>
<p>9. Train and save the model.</p>
<p>10. Predict data using the saved model.</p>
<p>11. Evaluate the created model and check its accuracy.</p>

In [0]:
# Created RDD FILE of autompg
autompg_f = spark.createDataFrame(autompg)
autompg_frdd=autompg_f.rdd
print("RDD File of linear regression")
autompg_frdd.take(8)

RDD File of linear regression
Out[199]: [Row(mpg=16.21463489067917, displacement=16.071718010941588, horsepower=14.587796515529897, weight=15.449113333404602, accelaration=11.601432303889403),
 Row(mpg=18.54372386038497, displacement=17.078059462108993, horsepower=18.57389896339249, weight=20.828405346898133, accelaration=12.750306538967422),
 Row(mpg=16.77708806854089, displacement=15.840512005319477, horsepower=17.734115478857674, weight=16.47878284460673, accelaration=12.351655202812314),
 Row(mpg=18.74553920823638, displacement=17.471583574034945, horsepower=18.825468608628046, weight=16.890875398111056, accelaration=11.359074567437721),
 Row(mpg=16.267082448487436, displacement=18.232842147603936, horsepower=19.373804038023327, weight=17.64422536415718, accelaration=11.244739799494399),
 Row(mpg=17.357269130855922, displacement=16.26378732259953, horsepower=13.740785753765167, weight=17.83249258360484, accelaration=11.527372617450187),
 Row(mpg=17.870668743114912, displacement=17.

In [0]:
autompg_frdd.take(8)

Out[214]: [Row(mpg=16.21463489067917, displacement=16.071718010941588, horsepower=14.587796515529897, weight=15.449113333404602, accelaration=11.601432303889403),
 Row(mpg=18.54372386038497, displacement=17.078059462108993, horsepower=18.57389896339249, weight=20.828405346898133, accelaration=12.750306538967422),
 Row(mpg=16.77708806854089, displacement=15.840512005319477, horsepower=17.734115478857674, weight=16.47878284460673, accelaration=12.351655202812314),
 Row(mpg=18.74553920823638, displacement=17.471583574034945, horsepower=18.825468608628046, weight=16.890875398111056, accelaration=11.359074567437721),
 Row(mpg=16.267082448487436, displacement=18.232842147603936, horsepower=19.373804038023327, weight=17.64422536415718, accelaration=11.244739799494399),
 Row(mpg=17.357269130855922, displacement=16.26378732259953, horsepower=13.740785753765167, weight=17.83249258360484, accelaration=11.527372617450187),
 Row(mpg=17.870668743114912, displacement=17.319751226905254, horsepower=15

<h2>Without Labeled Point</h2>

In [0]:
# Vectorization of Data of autompg
from pyspark.ml.feature import VectorAssembler
autompg_f = autompg_f.withColumnRenamed('mpg', "target")  

vectorAssembler = VectorAssembler(inputCols = ['displacement','horsepower','weight','accelaration'], outputCol = 'features')
autompg_f_df_assembled = vectorAssembler.transform(autompg_f)
autompg_f_df_assembled = autompg_f_df_assembled.select(['features', 'target'])
autompg_f_df_assembled.show(8, truncate=False)

+-----------------------------------------------------------------------------+------------------+
|features                                                                     |target            |
+-----------------------------------------------------------------------------+------------------+
|[16.071718010941588,14.587796515529897,15.449113333404602,11.601432303889403]|16.21463489067917 |
|[17.078059462108993,18.57389896339249,20.828405346898133,12.750306538967422] |18.54372386038497 |
|[15.840512005319477,17.734115478857674,16.47878284460673,12.351655202812314] |16.77708806854089 |
|[17.471583574034945,18.825468608628046,16.890875398111056,11.359074567437721]|18.74553920823638 |
|[18.232842147603936,19.373804038023327,17.64422536415718,11.244739799494399] |16.267082448487436|
|[16.26378732259953,13.740785753765167,17.83249258360484,11.527372617450187]  |17.357269130855922|
|[17.319751226905254,15.98975587329068,16.58859305727713,11.198982652661508]  |17.870668743114912|
|[15.83136

In [0]:
# Dividing the Dataset into training and test data
(train_data_auto, test_data_auto) = autompg_f_df_assembled.randomSplit([0.7, 0.3])
print("Train count:", train_data_auto.count())
print("Test count:", test_data_auto.count())

Train count: 350184
Test count: 149816


<b><H5>Ridge Regression Model</H5></b>

In [0]:
#Ridge Regression
ridgeregression=LinearRegression(featuresCol="features", labelCol="target", maxIter=100,regParam=0.1, elasticNetParam=0.0,aggregationDepth=2,fitIntercept=True, standardization=True)

# Training the Data
modelridge = ridgeregression.fit(train_data_auto)
print("Coefficients: " + str(modelridge.coefficients))
print("Intercept: " + str(modelridge.intercept))
print("")
print("Training Summary")

trainingSummaryridge = modelridge.summary
print("RMSE: %f" % trainingSummaryridge.rootMeanSquaredError)
print("r2: %f" % trainingSummaryridge.r2)
print("objectiveHistory: %s" % str(trainingSummaryridge.objectiveHistory))
trainingSummaryridge.residuals.show()

Coefficients: [-0.0007893362006116618,-0.001551828033630801,-0.002333319805935673,0.0021026735308392287]
Intercept: 16.853322208124244

Training Summary
RMSE: 1.302637
r2: 0.000011
objectiveHistory: [0.0]
+--------------------+
|           residuals|
+--------------------+
|  1.3162540557134221|
|  0.9713593494258461|
|  -1.114350446745334|
|  -0.735809687373866|
|   1.863134499621797|
|  0.5486279810727268|
|  2.5786359208073684|
| 0.03583227402123512|
|-0.12838204788238272|
| 0.42484537260191857|
| -2.1466207519932112|
|  0.7300117998307059|
|-0.10098172320146048|
|  1.0332846218551595|
| -0.7332565535238729|
| -1.0060887097741613|
| -1.7197607781349973|
|  1.2918859871151973|
|  0.6305985557546627|
| -1.4604297022315134|
+--------------------+
only showing top 20 rows



In [0]:
# For Deleteing  an Existing Path
dbutils.fs.rm('/tmp/mllib-persistence-example/modelridge', True)

# Saving an Model
from pyspark.ml.regression import LinearRegression
basePath = "/tmp/mllib-persistence-example"
modelridge.save(basePath + "/modelridge")

# Loading an Model
from pyspark.ml.regression import LinearRegressionModel
loaded_model_ridge=LinearRegressionModel.load('/tmp/mllib-persistence-example//modelridge')

In [0]:
#Prediction on Testing Data using Loaded Model and Evaluation (RIDGE MODEL)
from pyspark.ml.evaluation import RegressionEvaluator
evaluator = RegressionEvaluator(labelCol="target", predictionCol="prediction", metricName="rmse")

predictions_ridge = loaded_model_ridge.transform(test_data_auto)
predictions_ridge.select("features","target", "prediction").show(10)

rmse1 = evaluator.evaluate(predictions_ridge)
print("Ridge Regression RMSE:", rmse1)

+--------------------+------------------+------------------+
|            features|            target|        prediction|
+--------------------+------------------+------------------+
|[11.7658581543159...|19.076381714581373|16.803881458008362|
|[11.8487491568500...|17.689809830553816|16.801333241651538|
|[12.1064938281083...| 16.45926357370859|  16.7981472296484|
|[12.4647338332118...|18.197819446796597|16.802456075972408|
|[12.4834828170504...| 16.73531486529013|16.805017043356166|
|[12.5437746754105...|18.308694444762956| 16.80010118873976|
|[12.5563182512813...|19.238093265769336| 16.79845645696206|
|[12.6135033389889...|  18.6427846827988| 16.80437891256555|
|[12.6212441065453...|15.642577852189657|16.798373650299848|
|[12.6220822401825...|17.765693618944223| 16.80343680572395|
+--------------------+------------------+------------------+
only showing top 10 rows

Linear Regression RMSE: 1.3085275478864684


<b><H5>LASSO Regression Model</H5></b>

In [0]:
#LASSO Regression
lassoregression=LinearRegression(featuresCol="features", labelCol="target", maxIter=100,regParam=0.1, elasticNetParam=1,aggregationDepth=2,fitIntercept=True, standardization=True)

# Training the Data
modellasso = lassoregression.fit(train_data_auto)
print("Coefficients: " + str(modellasso.coefficients))
print("Intercept: " + str(modellasso.intercept))
print("")
print("Training Summary")

trainingSummarylasso = modellasso.summary
print("RMSE: %f" % trainingSummarylasso.rootMeanSquaredError)
print("r2: %f" % trainingSummarylasso.r2)
print("objectiveHistory: %s" % str(trainingSummarylasso.objectiveHistory))
trainingSummarylasso.residuals.show()

Coefficients: [0.0,0.0,0.0,0.0]
Intercept: 16.798768921970225

Training Summary
RMSE: 1.302644
r2: -0.000000
objectiveHistory: [0.5000000000000142]
+--------------------+
|           residuals|
+--------------------+
|  1.3181732753212394|
|  0.9687445747317653|
| -1.1041106703429904|
| -0.7370246995946168|
|  1.8678088916030546|
|  0.5464221970880487|
|     2.5835864295587|
| 0.04085494156954539|
|-0.13151037201525284|
|  0.4313942638062471|
| -2.1404666943838855|
|  0.7328824740250433|
|-0.09795287740022118|
|  1.0388740467447768|
| -0.7279129658077963|
| -0.9991805470781578|
|  -1.712805668362293|
|   1.290033497968622|
|  0.6344777161866162|
| -1.4558885092824827|
+--------------------+
only showing top 20 rows



In [0]:
# For Deleteing  an Existing Path
dbutils.fs.rm('/tmp/mllib-persistence-example/modellasso', True)

# Saving an Model
from pyspark.ml.regression import LinearRegression
basePath = "/tmp/mllib-persistence-example"
modellasso.save(basePath + "/modellasso")

# Loading an Model
from pyspark.ml.regression import LinearRegressionModel
loaded_model_lasso=LinearRegressionModel.load('/tmp/mllib-persistence-example//modellasso')

In [0]:
#Prediction on Testing Data using Loaded Model and Evaluation (RIDGE MODEL)
from pyspark.ml.evaluation import RegressionEvaluator
evaluator = RegressionEvaluator(labelCol="target", predictionCol="prediction", metricName="rmse")

predictions_lasso = loaded_model_lasso.transform(test_data_auto)
predictions_lasso.select("features","target", "prediction").show(10)

rmse1 = evaluator.evaluate(predictions_lasso)
print("Lasso Regression RMSE:", rmse1)

+--------------------+------------------+------------------+
|            features|            target|        prediction|
+--------------------+------------------+------------------+
|[11.7658581543159...|19.076381714581373|16.798768921970225|
|[11.8487491568500...|17.689809830553816|16.798768921970225|
|[12.1064938281083...| 16.45926357370859|16.798768921970225|
|[12.4647338332118...|18.197819446796597|16.798768921970225|
|[12.4834828170504...| 16.73531486529013|16.798768921970225|
|[12.5437746754105...|18.308694444762956|16.798768921970225|
|[12.5563182512813...|19.238093265769336|16.798768921970225|
|[12.6135033389889...|  18.6427846827988|16.798768921970225|
|[12.6212441065453...|15.642577852189657|16.798768921970225|
|[12.6220822401825...|17.765693618944223|16.798768921970225|
+--------------------+------------------+------------------+
only showing top 10 rows

Linear Regression RMSE: 1.3085237965927903


<h2>With Labeled Point</h2>

In [0]:
#test train split
split = autompg_frdd.randomSplit([0.7,0.3],seed=1)
train_data = split[0]
test_data = split[1]

In [0]:
from pyspark.mllib.regression import LabeledPoint

Labeled_Point_train = train_data.map(lambda data : LabeledPoint(data[0],data[1:5]))
Labeled_Point_test = test_data.map(lambda data : LabeledPoint(data[0],data[1:5]))

<h5>Ridge Regression Model

In [0]:
from pyspark.mllib.regression import RidgeRegressionWithSGD as ridgeSGD
ridgemodel_LP = ridgeSGD.train(data = Labeled_Point_train,iterations = 100,step = 0.0000005,regParam = 0.05,intercept = True)

In [0]:
print("Intercept:")
print(ridgemodel.intercept)
print("Coefficients of the Variables:")
print("displacement:",ridgemodel_LP.weights[0])
print("horsepower:",ridgemodel_LP.weights[1])
print("weight:",ridgemodel_LP.weights[2])
print("accelaration:",ridgemodel_LP.weights[3])

Intercept:
1.0000134397362521
Coefficients of the Variables:
displacement: 0.0002264823400272479
horsepower: 0.0002264723164517805
weight: 0.00022648274921181972
accelaration: 0.00015369328230565015


In [0]:
# For Deleteing  an Existing Path
dbutils.fs.rm('/home/ridgemodel_LP', True)

# Saving an Model
ridgemodel_LP.save(sc, '/home/ridgemodel_LP')

# Loading an Model
from pyspark.mllib.regression import RidgeRegressionModel as RidgeModel
Loaded_ridgemodel_LP = RidgeModel.load(sc, '/home/ridgemodel_LP')

# Predicting
predictedridge = Labeled_Point_test.map(lambda data : (float(data.label) , float(Loaded_ridgemodel_LP.predict(data.features))))

In [0]:
predictedridge.take(5)

Out[235]: [(18.54372386038497, 1.0147647031342415),
 (16.267082448487436, 1.0142548400438725),
 (17.433688135617153, 1.0122797362521128),
 (18.170823773167413, 1.0133403405725354),
 (16.755450000092626, 1.0132305608136987)]

In [0]:
#RMSE Value
from pyspark.mllib.evaluation import RegressionMetrics as metrics
error = metrics(predictedridge)
print("RMSE:")
print(error.rootMeanSquaredError)

# R squared value
print("R Squared Value")
print(error.r2)

RMSE:
15.841844021728894
R Squared Value
-926235571.1464496


<h5>LASSO Regression Model

In [0]:
from pyspark.mllib.regression import LassoWithSGD as lassoSGD
lasso_model_LP = lassoSGD.train(data = Labeled_Point_train,iterations = 100, step = 0.0000005,regParam = 0.05, intercept = True)

In [0]:
print("Intercept:")
print(lasso_model_LP.intercept)
print("Coefficients of the Variables:")
print("displacement:",lasso_model_LP.weights[0])
print("horsepower:",lasso_model_LP.weights[1])
print("weight:",lasso_model_LP.weights[2])
print("accelaration:",lasso_model_LP.weights[3])

Intercept:
1.0000134393740023
Coefficients of the Variables:
displacement: 0.00022643967389410704
horsepower: 0.00022642965031810115
weight: 0.0002264400830785971
accelaration: 0.00015365061245824254


In [0]:
# For Deleteing  an Existing Path
dbutils.fs.rm('/home/lassomodel_LP', True)

# Saving an Model
lasso_model_LP.save(sc, '/home/lassomodel_LP')

# Loading an Model
from pyspark.mllib.regression import LassoModel as LassoModel
Loaded_lassomodel_LP = LassoModel.load(sc, '/home/lassomodel_LP')

# Predicting
predictedlasso = Labeled_Point_test.map(lambda data : (float(data.label) , float(Loaded_lassomodel_LP.predict(data.features))))

In [0]:
predictedlasso.take(5)

Out[253]: [(18.54372386038497, 1.0147617492825602),
 (16.267082448487436, 1.0142520028921727),
 (17.433688135617153, 1.0122772649141767),
 (18.170823773167413, 1.0133376722543916),
 (16.755450000092626, 1.0132279080925348)]

In [0]:
#RMSE Value
from pyspark.mllib.evaluation import RegressionMetrics as metrics
error = metrics(predictedlasso)
print("RMSE:")
print(error.rootMeanSquaredError)

# R squared value
print("R Squared Value")
print(error.r2)

RMSE:
15.841846649183495
R Squared Value
-926591166.9328218


<h1><p>Problem 4</h1></p>
<h4>Compare the lasso regression and ridge regression models and provide your commentary.<h4>

Lasso regression is a type of linear regression that uses shrinkage. Shrinkage is where data values are shrunk towards a central
point, like the mean.

Ridge regression is a method of estimating the coefficients of multiple-regression models in scenarios where linearly independent variables are highly correlated.

The main difference between Ridge and LASSO Regression is that if ridge regression can shrink the coefficient close to 0 so that all predictor variables are retained. Whereas LASSO can shrink the coefficient to exactly 0 so that LASSO can select and discard the predictor variables that have the right coefficient of 0.

During training, Lasso introduced a new hyperparameter, alpha, the coefficient to penalize weights. Ridge takes a step further and penalizes the model for the sum of squared
value of the weights.

Calculated RMSE Errors (Without Labelled Point)
               
Ridge Regression	1.302637 	
Lasso Regression	1.302644	

Calculated RMSE Errors and R Squared Values (Without Labelled Point)
Ridge Regression	15.841844021728894	
Lasso Regression	15.841846649183495	

From Errors we can interpret that we are getting similar results in both the type of regression. Also, RMSE is very in both the data from the models has been created randomlly following the normal distribution.